docs/docs/apis/focusutils.md

```diff
@@ -0,0 +1,78 @@
+---
+id: apis/focusutils
+title: FocusUtils
+layout: docs
+category: Interfaces
+permalink: docs/apis/focusutils.html
+next: apis/input
+---
+
+This interface provides methods related to managing focusable components.
+
+## Methods
+``` javascript
+// Allows to register a component for being automatically focused during the
+// current mount cycle.
+// The same method which is used internally to handle autoFocus property for
+// the components which might support being focused (View, Button, etc.)
+
+// Moving the focus manually from the application code might lead to a complex
+// logic on the application level and to race conditions when several components
+// want to be focused on mount at the same time. Those race conditions might
+// result in an unreliable behaviour, especially for the screen reader users.
+// For example, when some element inside the dialog is focused, the screen reader
+// announces the dialog, but only when nothing inside the dialog was focused before,
+// and if the focus is moved inside the dialog, the second focused element
+// might interrupt the dialog announcement and the screen reader users won't be
+// able to understand where they are.
+// Another thing is that it is often required to have different autofocusing logic
+// on different platforms. For example it is often ok to automatically focus the
+// input field on desktop when something shows, but on mobile it will bring up the
+// keyboard which is not always the desired behaviour.
+
+// To tackle above we have this concept of smart automatic focus.
+// See the Types.AutoFocus enum definition:
+
+// export enum AutoFocus {
+//     No = 0,
+//     Yes = 1, // All platforms, any state of the keyboard navigation mode, default
+//              // priority, no delay.
+
+//     // We can target the keyboard navigation mode (if none of
+//     // WhenNavigatingWithKeyboard and WhenNavigatingWithoutKeyboard are specified,
+//     // both are assumed to be enabled).
+//     WhenNavigatingWithKeyboard,
+//     WhenNavigatingWithoutKeyboard,
+
+//     // We can target the platform (if none of the platforms are specified, all
+//     // platforms are assumed to be enabled).
+//     Android,
+//     IOS,
+//     Web,
+//     Windows,
+//     Mac,
+
+//     // Sometimes a common high level component has default autofocusable (for
+//     // example, close button in a dialog), but the subcomponents want to be
+//     // autofocused too (for example, some input inside a particular kind of dialog).
+//     // We can specify the priority (Low for the close button, High for the
+//     // input) without forking the logic on the application level (the input will
+//     // win if present, otherwise the close button will be focused).
+//     // Default priority is PriorityLow.
+//     PriorityLow,
```

Comment:
I don't like this way of exposing the concept. Given that you have only three discrete priority values, it's not clear how they compose. Also, components need to know about the values passed by other components, which is not a good design.

I assume that these conflicts don't occur very often. When they do, can't we apply a simple heuristic — like use the one that's "deeper" in the view hierarchy? If that's not possible, it would probably be better to use a numeric value rather than two or three discrete priorities. This is how, for example, zIndex and taxIndex work.